### ISSUE TITLE
"Overflow encountered when expanding vector" during solving

### ISSUE THREAD

**Issue by PhilippWendler** [Z3Prover/z3#864]
I have several queries where Z3 aborts with `Overflow encountered when expanding vector`. In some cases this error comes after only a few seconds for not-too-big formulas. I am currently using Z3 4.5.0-124-gc1480b4, but I can reproduce this with the latest git revision.

So I am wondering whether
- this might be a bug that leads to unexpectedly high memory usage?
- something could be optimized to solve such formulas with less memory? The query actually does not seem to be that difficult: If I serialize the solver and give the resulting SMTLIB2 formula to Z3, it can be solved instantly.
- Z3 could be changed to allow the use of larger vectors? I have a 64bit system and would be fine with Z3 taking a few more gigabytes for solving the formula (I am currently using a memory limit of 15 GB, and Z3 just uses 4 and then gives up).

The queries where I see this problem all use linear arithmetic over Ints and UFs, some of them also use arrays and/or quantifiers. Interestingly, I have not encountered this problem for bitvector queries so far.
I cannot remember seeing this problem earlier, it looks like it is related to a change in Z3 during the last months?

Here is an [archive](http://students.fim.uni-passau.de/~wendler/z3-overflow-encountered.tar.bz2) with a replay log and the same query as serialized solver output for one of my problems. If you need more information, just tell me.

**Comment by NikolajBjorner** [comment]
Thanks, there seem to be some related entries on this.
The stack I see is:

 	z3.exe!memory::reallocate(void * p, unsigned int s) Line 304	C++
 	z3.exe!vector<smt::conflict_resolution::tp_elem,0,unsigned int>::expand_vector() Line 77	C++
 	z3.exe!vector<smt::conflict_resolution::tp_elem,0,unsigned int>::push_back(const smt::conflict_resolution::tp_elem & elem) Line 284	C++
 	z3.exe!smt::conflict_resolution::get_proof(smt::justification * js) Line 1056	C++
 	z3.exe!smt::conflict_resolution::visit_b_justification(smt::literal l, smt::b_justification js) Line 1084	C++
>	z3.exe!smt::conflict_resolution::mk_conflict_proof(smt::b_justification conflict, smt::literal not_l) Line 1295	C++
 	z3.exe!smt::conflict_resolution::finalize_resolve(smt::b_justification conflict, smt::literal not_l) Line 488	C++
 	z3.exe!smt::conflict_resolution::resolve(smt::b_justification conflict, smt::literal not_l) Line 598	C++

In other words, you have proofs turned on and as it reconstruct proof objects it runs out of memory.
It is possible that proof objects are getting impossibly large and there is no way around it, but another thing to investigate is whether proof reconstruction is missing out of shared sub-expressions and ends up unfolding a DAG needlessly.

**Comment by NikolajBjorner** [comment]
It seems to be a bug in tracking justifications. It ends up with a justification of the form:

p22 : clause (not (= id1@3 id4@3))
(or (= id1@3 id4@3)
    (= id2@3 p1_old@7)
    (= id2@3 |node1::__CPAchecker_TMP_0@5|)
    (= |node1::__CPAchecker_TMP_0@5| p1_old@7)
    (= |node1::__CPAchecker_TMP_0@5| id2@3)
    (not (= p4_new@7 nomsg@2))
    (not (= p3_new@7 nomsg@2))
    (not (= p2_new@7 nomsg@2))
    (not (= p1_new@7 nomsg@2))
    (= nomsg@2 id2@3)
    (not (= p2_old@7 nomsg@2))
    (not (= id2@3 |node2::__CPAchecker_TMP_0@5|))
    (= id2@3 p2_old@7)
    (= id3@3 p3_old@7)
    (= id1@3 p4_old@7)
    (= id1@3 nomsg@2)
    (= id1@3 |node4::__CPAchecker_TMP_0@5|)
    (not (= |node4::__CPAchecker_TMP_0@5| id4@3))
    (not (<= r1@5 1)))

There is a discrepancy between the polarity of the literal that gets implied by the clause. 
So somewhere the invariant that justifications on Boolean variables are consistent with the literal assignment is broken. So the memory overflow is just a symptom.

**Comment by NikolajBjorner** [comment]
Try out the fix. The API changed so old replay logs will no longer work.

**Comment by PhilippWendler** [comment]
Thanks, I will test it, re-running all benchmarks might take a few days though.

Btw., in this situation I don't even need proofs, they are just enabled because AFAIK we can only enable them per context, not per solver. If we could disable proofs and model generation per solver this would probably be more efficient for us.

**Comment by PhilippWendler** [comment]
Hm, this commit breaks compilation with both gcc 4.8 and 5.4:
```
../src/smt/smt_context.cpp: In member function ‘bool smt::context::propagate_atoms()’:
../src/smt/smt_context.cpp:1416:33: error: expected unqualified-id before ‘not_eq’ token
                         literal not_eq = literal(l.var(), true);
                                 ^
../src/smt/smt_context.cpp:1417:135: error: expected primary-expression before ‘not_eq’ token
                         set_conflict(b_justification(mk_justification(eq_propagation_justification(get_enode(lhs), get_enode(rhs)))), not_eq);                        
                                                                                                                                       ^
../src/smt/smt_context.cpp:1417:141: error: expected primary-expression before ‘)’ token
                         set_conflict(b_justification(mk_justification(eq_propagation_justification(get_enode(lhs), get_enode(rhs)))), not_eq); 
```

**Comment by wintersteiger** [comment]
(Compilation fixed.)

**Comment by wintersteiger** [comment]
See also #864; the assertion failure there may be a root cause and possibly happens before the bogus justification makes it into the state.

**Comment by NikolajBjorner** [comment]
Closing...

**Comment by PhilippWendler** [comment]
I can now confirm the fix works.
